planner/function_dependency/fast_int_set.go

```diff
@@ -0,0 +1,376 @@
+	for i, ok := s.Next(0); ok; i, ok = s.Next(i + 1) {
+		large.Insert(i)
+	}
+	return large
+}
+
+func (s FastIntSet) Next(startVal int) (int, bool) {
+	if startVal < smallCutOff {
+		if startVal < 0 {
+			startVal = 0
+		}
+		// x=0, gap=64, which means there is no `1` after right shift.
+		if gap := bits.TrailingZeros64(s.small >> uint64(startVal)); gap < 64 {
+			return gap + startVal, true
+		}
+	}
+	if s.large != nil {
+		res := s.large.LowerBound(startVal)
+		return res, res != intsets.MaxInt
+	}
+	return intsets.MaxInt, false
+}
+
+// Remove is used to remove a value from the set. Nothing done if the value is not in the set.
+func (s *FastIntSet) Remove(i int) {
+	if i >= 0 && i < smallCutOff {
+		s.small &^= 1 << uint64(i)
+	}
+	if s.large != nil {
+		s.large.Remove(i)
+	}
+}
+
+// Clear is used to clear a fastIntSet and reuse it as an empty one.
+func (s *FastIntSet) Clear() {
+	s.small = 0
+	if s.large != nil {
+		s.large.Clear()
+	}
+}
+
+// Has is used ot judge whether a value is in the set.
+func (s FastIntSet) Has(i int) bool {
+	if i >= 0 && i < smallCutOff {
+		return (s.small & (1 << uint64(i))) != 0
+	}
+	if s.large != nil {
+		return s.large.Has(i)
+	}
+	return false
+}
+
+// IsEmpty is used to judge whether the int-set is empty.
+func (s FastIntSet) IsEmpty() bool {
+	return s.small == 0 && (s.large == nil || s.large.IsEmpty())
+}
+
+// SortedArray is used to return the in array of the set.
+func (s FastIntSet) SortedArray() []int {
+	if s.IsEmpty() {
+		return nil
+	}
+	if s.large != nil {
+		return s.large.AppendTo([]int(nil))
+	}
+	res := make([]int, 0, s.Len())
+	s.ForEach(func(i int) {
+		res = append(res, i)
+	})
+	return res
+}
+
+// ForEach call a function for each value in the int-set. (Ascend)
+func (s FastIntSet) ForEach(f func(i int)) {
+	if s.large != nil {
+		for x := s.large.Min(); x != intsets.MaxInt; x = s.large.LowerBound(x + 1) {
+			f(x)
+		}
+		return
+	}
+	for v := s.small; v != 0; {
+		// from the left to right.
+		i := bits.TrailingZeros64(v)
+		f(i)
+		v &^= 1 << uint(i)
+	}
+}
+
+// Copy returns a copy of s.
+func (s FastIntSet) Copy() FastIntSet {
+	c := FastIntSet{}
+	c.small = s.small
+	if s.large != nil {
+		c.large = new(intsets.Sparse)
+		c.large.Copy(s.large)
+	}
+	return c
+}
+
+// CopyFrom clear the receiver to be a copy of the param.
+func (s *FastIntSet) CopyFrom(target FastIntSet) {
+	s.small = target.small
+	if target.large != nil {
+		if s.large == nil {
+			s.large = new(intsets.Sparse)
+		}
+		s.large.Copy(target.large)
+	} else {
+		if s.large != nil {
+			s.large.Clear()
+		}
+	}
+}
+
+// Equals returns whether two int-set are identical.
+func (s FastIntSet) Equals(rhs FastIntSet) bool {
+	if s.large == nil && rhs.large == nil {
+		return s.small == rhs.small
+	}
+	if s.large != nil && rhs.large != nil {
+		return s.large.Equals(rhs.large)
+	}
+	// how come to this? eg: a set operates like: +1, +65, -65, resulting a large int-set with only small numbers.
```

Comment:
The example here is wrong? If we insert a negative number, there will be a large part.

Reply:
sorry, here means, add two number, 1, 65, then the 65 is removed

Reply:
Oh, i got it.